src/widgets/scrollbar.rs

```diff
@@ -0,0 +1,853 @@
+            return true;
+        }
+        false
+    }
+
+    fn get_track_start_end(&self, area: Rect) -> (u16, u16, u16) {
+        match self.orientation {
+            ScrollbarOrientation::VerticalRight => {
+                (area.top(), area.bottom(), area.right().saturating_sub(1))
+            }
+            ScrollbarOrientation::VerticalLeft => (area.top(), area.bottom(), area.left()),
+            ScrollbarOrientation::HorizontalBottom => {
+                (area.left(), area.right(), area.bottom().saturating_sub(1))
+            }
+            ScrollbarOrientation::HorizontalTop => (area.left(), area.right(), area.top()),
+        }
+    }
+
+    fn get_thumb_start_end(
+        &self,
+        state_content_length: u16,
+        state_position: u16,
+        viewport_content_length: u16,
+        track_start: u16,
+        track_end: u16,
+    ) -> (u16, u16) {
+        let scroll_position_ratio = (state_position as f64 / state_content_length as f64).min(1.0);
+
+        let thumb_size = (((viewport_content_length as f64 / state_content_length as f64)
+            * (track_end - track_start) as f64)
+            .round() as u16)
+            .max(1);
+
+        let maximum_main_size = (track_end - track_start).saturating_sub(thumb_size);
+
+        let thumb_start =
+            track_start + (scroll_position_ratio * maximum_main_size as f64).round() as u16;
+        let thumb_end = thumb_start + thumb_size;
+
+        (thumb_start, thumb_end)
+    }
+}
+
+impl<'a> StatefulWidget for Scrollbar<'a> {
+    type State = ScrollbarState;
+
+    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
+        //
+        // For ScrollbarOrientation::VerticalRight
+        //
+        //                   ┌───────── track_axis  (x)
+        //                   v
+        //   ┌───────────────┐
+        //   │               ║<──────── track_start (y1)
+        //   │               █
+        //   │               █
+        //   │               ║
+        //   │               ║<──────── track_end   (y2)
+        //   └───────────────┘
+        //
+        // For ScrollbarOrientation::HorizontalBottom
+        //
+        //   ┌───────────────┐
+        //   │               │
+        //   │               │
+        //   │               │
+        //   └═══███═════════┘<──────── track_axis  (y)
+        //    ^             ^
+        //    │             └────────── track_end   (x2)
+        //    │
+        //    └──────────────────────── track_start (x1)
+        //
+
+        // Find track_start, track_end, and track_axis
+        let area = self.get_track_area(area);
+        let (track_start, track_end, track_axis) = self.get_track_start_end(area);
+
+        if self.should_not_render(track_start, track_end, state.content_length) {
+            return;
+        }
+
+        let (thumb_start, thumb_end) = self.get_thumb_start_end(
+            state.content_length,
+            state.position,
```

Comment:
```suggestion
        let (thumb_start, thumb_end) = self.get_thumb_start_end(
            &state
```

(if the previous suggestion is applied)

Reply:
Resolved in https://github.com/tui-rs-revival/ratatui/pull/228/commits/2284b202bd71c2e1c0549aaa37a89276aeee2c42.